### ISSUE TITLE
Crash for UIDiffableDataSourceUpdate duplicate identifiers: Client.HomepageDiffableDataSource.HomeItem.topSite(Client.TopSiteState)

### ISSUE THREAD

**Issue by data-sync-user** [mozilla-mobile/firefox-ios#23056]
I saw this crash when I tried to run the app on `main` this morning after the long weekend.

It happened every time I launch the app until I removed the app from my simulator. :thinking:  So it seems to be some sort of bad state. Possibly something related to other work I was doing on a branch, but it might bear some investigation…

```
Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Fatal: supplied item identifiers are not unique. Duplicate identifiers: {(
    Client.HomepageDiffableDataSource.HomeItem.topSite(Client.TopSiteState)
)}'
```

I had the toolbar refactor, menu refactor, and unified search feature flags enabled. Not sure if related.

Stack Trace:

```
2024-11-12 09:16:28.432 💙 INFO [redux] Store - Dispatched action: ToolbarAction websiteLoadingStateDidChange
2024-11-12 09:16:28.436 💙 INFO [redux] Store - Dispatched action: ToolbarAction urlDidChange
2024-11-12 09:16:28.437 💙 INFO [redux] Store - Dispatched action: ToolbarMiddlewareAction urlDidChange
2024-11-12 09:16:28.437 💙 INFO [redux] Store - Dispatched action: ToolbarAction borderPositionChanged
2024-11-12 09:16:28.439 💙 INFO [redux] Store - Dispatched action: MicrosurveyPromptAction showPrompt
2024-11-12 09:16:28.440 💙 INFO [redux] Store - Dispatched action: FakespotAction tabDidChange
2024-11-12 09:16:28.441 💙 INFO [redux] Store - Dispatched action: FakespotAction setAppearanceTo
2024-11-12 09:16:28.460 💙 INFO [redux] Store - Dispatched action: ToolbarAction websiteLoadingStateDidChange
nw_connection_copy_connected_local_endpoint_block_invoke [C5] Client called nw_connection_copy_connected_local_endpoint on unconnected nw_connection
nw_connection_copy_connected_remote_endpoint_block_invoke [C5] Client called nw_connection_copy_connected_remote_endpoint on unconnected nw_connection
nw_connection_copy_protocol_metadata_internal_block_invoke [C5] Client called nw_connection_copy_protocol_metadata_internal on unconnected nw_connection
AX Safe category class 'SLHighlightDisambiguationPillViewAccessibility' was not found!
2024-11-12 09:16:29.803 💙 INFO [redux] Store - Dispatched action: TopSitesAction retrievedUpdatedSites
2024-11-12 09:16:29.804 💙 INFO [redux] Store - Dispatched action: TopSitesAction retrievedUpdatedSites
*** Assertion failure in -[_UIDiffableDataSourceUpdate initWithIdentifiers:sectionIdentifiers:action:desinationIdentifier:relativePosition:destinationIsSection:], _UIDiffableDataSourceHelpers.m:504
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Fatal: supplied item identifiers are not unique. Duplicate identifiers: {(
    Client.HomepageDiffableDataSource.HomeItem.topSite(Client.TopSiteState)
)}'
*** First throw call stack:
(
	0   CoreFoundation                      0x00000001804ae0f8 __exceptionPreprocess + 172
	1   libobjc.A.dylib                     0x0000000180087db4 objc_exception_throw + 56
	2   Foundation                          0x0000000180d17058 _userInfoForFileAndLine + 0
	3   UIKitCore                           0x000000018586a728 -[_UIDiffableDataSourceUpdate initWithIdentifiers:sectionIdentifiers:action:desinationIdentifier:relativePosition:destinationIsSection:] + 512
	4   UIKitCore                           0x000000018586aa68 -[_UIDiffableDataSourceUpdate initWithItemIdentifiers:appendingToDestinationSectionIdentifier:] + 88
	5   UIKitCore                           0x0000000184e4f0c8 -[__UIDiffableDataSourceSnapshot appendItemsWithIdentifiers:intoSectionWithIdentifier:] + 124
	6   UIKitCore                           0x000000018487412c __swift_destroy_boxed_opaque_existential_1Tm + 12740
	7   Client                              0x0000000105648070 $s6Client26HomepageDiffableDataSourceC20applyInitialSnapshot5stateyAA0B5StateV_tF + 820
	8   Client                              0x0000000104b9fa68 $s6Client22HomepageViewControllerC8newState5stateyAA0bF0V_tF + 272
	9   Client                              0x0000000104ba3938 $s6Client22HomepageViewControllerC5Redux15StoreSubscriberAadEP8newState5statey0gI4TypeQz_tFTW + 12
	10  Client                              0x0000000105bcfcb4 $s5Redux15StoreSubscriberPAAE8newState5stateyyp_tF + 308
	11  Client                              0x0000000104ba3994 $s6Client22HomepageViewControllerC5Redux18AnyStoreSubscriberAadEP8newState5stateyyp_tFTW + 44
	12  Client                              0x0000000105bd0208 $s5Redux19SubscriptionWrapperC08originalB0011transformedB010subscriberACyxGAA0B0CyxG_AIyqd__GSgAA18AnyStoreSubscriber_ptclufcyqd__Sg_qd__tcfU_ + 160
	13  Client                              0x0000000105bd0760 $s5Redux12SubscriptionC9newValues8oldState0cF0yxSg_xtF + 76
	14  Client                              0x0000000105bd0a94 $s5Redux12SubscriptionC6selectyACyqd__Gqd__xclFyyqd__Sg_qd__tccfU_yxSg_xtcfU_ + 364
	15  Client                              0x0000000105bd0760 $s5Redux12SubscriptionC9newValues8oldState0cF0yxSg_xtF + 76
	16  Client                              0x0000000105bd03a4 $s5Redux19SubscriptionWrapperC9newValues8oldState0dG0yx_xtF + 172
	17  Client                              0x0000000105bcd858 $s5Redux5StoreC5statexvWyAA19SubscriptionWrapperCyxGXEfU_ + 164
	18  Client                              0x0000000105bcfb4c $s5Redux5StoreC5statexvWyAA19SubscriptionWrapperCyxGXEfU_TA + 16
	19  libswiftCore.dylib                  0x000000019307e148 $sSTsE7forEachyyy7ElementQzKXEKF + 440
	20  Client                              0x0000000105bcd798 $s5Redux5StoreC5statexvW + 196
	21  Client                              0x0000000105bcf408 $s5Redux5StoreC5statexvsTf4gn_n + 168
	22  Client                              0x0000000105bcd990 $s5Redux5StoreC5statexvs + 28
	23  Client                              0x0000000105bce970 $s5Redux5StoreC13executeAction33_5E728F9C6CCF0FC19C907B7493EFDCB5LLyyAA0D0CF + 380
	24  Client                              0x0000000105bce6e8 $s5Redux5StoreC20processQueuedActions33_5E728F9C6CCF0FC19C907B7493EFDCB5LLyyF + 324
	25  Client                              0x0000000105bce320 $s5Redux5StoreC8dispatchyyAA6ActionCF + 524
	26  Client                              0x0000000105bce560 $s5Redux5StoreC8dispatchyyAA6ActionCFyyYbcfU_ + 92
	27  Client                              0x0000000105bce594 $sIegh_IeyBh_TR + 28
	28  libdispatch.dylib                   0x000000010c17fec4 _dispatch_call_block_and_release + 24
	29  libdispatch.dylib                   0x000000010c18173c _dispatch_client_callout + 16
	30  libdispatch.dylib                   0x000000010c1913f8 _dispatch_main_queue_drain + 1228
	31  libdispatch.dylib                   0x000000010c190f1c _dispatch_main_queue_callback_4CF + 40
	32  CoreFoundation                      0x000000018040e960 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 12
	33  CoreFoundation                      0x0000000180409078 __CFRunLoopRun + 1936
	34  CoreFoundation                      0x00000001804084d4 CFRunLoopRunSpecific + 572
	35  GraphicsServices                    0x000000018ef2aae4 GSEventRunModal + 160
	36  UIKitCore                           0x00000001853d0a28 -[UIApplication _run] + 868
	37  UIKitCore                           0x00000001853d46b0 UIApplicationMain + 124
	38  Client                              0x000000010545e4a4 main + 924
	39  dyld                                0x000000010ac45544 start_sim + 20
	40  ???                                 0x000000010ae0e154 0x0 + 4477477204
	41  ???                                 0x2e03800000000000 0x0 + 3315634488163172352
)
libc++abi: terminating due to uncaught exception of type NSException
```

See attached screenshot.

┆Issue is synchronized with this [Jira Bug](https://mozilla-hub.atlassian.net/browse/FXIOS-10518)

**Comment by data-sync-user** [comment]
➤ Isabella commented:

Thanks Cyndi Chin ! I think I accidentally toggled your feature flag in the debug settings so that would explain this. 😀

**Comment by data-sync-user** [comment]
➤ Cyndi Chin commented:

This issue is due to the top sites logic being incomplete and this was intentional to keep PRs small. This PR ( https://github.com/mozilla-mobile/firefox-ios/pull/23064 ) has been merged and should fix the issue. The issue is that we were not removing duplicates in the list of top sites, which causes an issues with diffable datasource because the item identifiers were no longer unique causing the API to not be able to properly do automatic diffing. Since there may be some cases where the list may not contain unique items, I created another ticket here to further safeguard our diffable datasource code: https://mozilla-hub.atlassian.net/browse/FXIOS-10540 ( https://mozilla-hub.atlassian.net/browse/FXIOS-10540|smart-link )
